EndlessClient/Network/UnknownEntitiesRequester.cs

```diff
@@ -0,0 +1,104 @@
+﻿using EndlessClient.GameExecution;
+using EOLib.Domain.Map;
+using EOLib.Net;
+using EOLib.Net.Communication;
+using Microsoft.Xna.Framework;
+using System;
+
+namespace EndlessClient.Network
+{
+    public class UnknownEntitiesRequester : GameComponent
+    {
+        private readonly ICurrentMapStateRepository _currentMapStateRepository;
+        private readonly IPacketSendService _packetSendService;
+        private DateTime _lastRequestTime;
+        private const double REQUEST_INTERVAL_SECONDS = 1;
+
+        public UnknownEntitiesRequester(IEndlessGameProvider gameProvider,
+                                      ICurrentMapStateRepository currentMapStateRepository,
+                                      IPacketSendService packetSendService)
+            : base((Game) gameProvider.Game)
+        {
+            _currentMapStateRepository = currentMapStateRepository;
+            _packetSendService = packetSendService;
+            _lastRequestTime = DateTime.Now;
+        }
+
+        public override void Update(GameTime gameTime)
+        {
+            if ((DateTime.Now - _lastRequestTime).TotalSeconds >= REQUEST_INTERVAL_SECONDS)
+            {
+                IPacket request = null;
+                if (_currentMapStateRepository.UnknownNPCIndexes.Count > 0 && _currentMapStateRepository.UnknownPlayerIDs.Count > 0)
+                {
+                    request = CreateRequestForBoth();
+                }
+                else if (_currentMapStateRepository.UnknownNPCIndexes.Count > 0)
+                {
+                    request = CreateRequestForNPCs();
+                }
+                else if (_currentMapStateRepository.UnknownPlayerIDs.Count > 0)
+                {
+                    request = CreateRequestForPlayers();
+                }
+
+                try
+                {
+                    if (request != null)
+                        _packetSendService.SendPacket(request);
+                }
+                catch (NoDataSentException)
+                {
+                    // do nothing.. should we log these?
```

Comment:
My error handling model is pretty bad (or just largely incomplete). I throw these errors but really all they signal is "uh oh time to return to the main menu". Two options:
1. Leave this exception unhandled, which will cause the game to go to the main menu if the exception is ever thrown (pretty sure there's a higher level try block somewhere that does this...)
2. Swallow it and move the clearing of the collections that's currently in the `finally` block to the `try` block, so the next pass will try it again

I'd lean more toward option 2 because this shouldn't be a fatal error; any unknown IDs will keep collecting until the next attempt, and if there's really a problem with the networking stack something else will explode and send everything back to the main menu (or just crash the game lol)